src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -587,7 +588,7 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
                     }
                 }
             }
-            else if (Sse2.IsSupported)
+            else if (Sse2.IsSupported || WasmBase.IsSupported)
```

Comment:
If the `WASM` hookups also existed for the `Vector128<T>` APIs, then this could be `Vector128.IsHardwareAccelerated`

and then the line below could just be `int matches = Vector128.CompareEqual(values, search).ExtractMostSignificantBits();`

The code would then work for x86, x64, Arm64, and WASM without needing to have many separate paths.

Reply:
The changes here to SpanHelper's are to demonstrate the `WasmBase` functionality and improve Span performance on wasm. They are quite small and done in similar way as other platforms here, so I hope they don't make the situation worse.

It will be nice to make these helpers use Vector128<T> for all the platforms you mentioned. Lets do that in separate PR? I can help with it.

For this PR I aim to get partial WasmBase implementation and speedup the Span on wasm. I am keeping the WasmBase internal for now, so that I can make changes if something doesn't go as desired. I plan to get it to API review once it is complete. I didn't propose any API for review yet, so hope this is good plan.

Reply:
> It will be nice to make these helpers use Vector128<T> for all the platforms you mentioned. Lets do that in separate PR?

Sounds good. Like I mentioned below, this is what we're hoping to do for more paths overall since it's less code to support more platforms and with the same perf for most scenarios.

>  am keeping the WasmBase internal for now, so that I can make changes if something doesn't go as desired. I plan to get it to API review once it is complete.

Yep, that's definitely good since we can't expose any new public API without it going through review first.

> I didn't propose any API for review yet, so hope this is good plan.

https://github.com/dotnet/runtime/issues/53730 should be the existing API proposal, so we shouldn't need anything new. Just update this one with any tweaks necessary before we take it to review.